File: JavaScript/4_SortingAlgorithms/6_RadixSort.js
/*

Time Complexity => O(kn)  k => maxdigits and n => number of elems in array
Space Complexity => O(n+k)

Radix Sort is a fun Sorting Algorithm

Uses the digits to sort the array

for base 10 numbers(decimals) we have 10 buckets from (0 to 9)

we find max number of digits from the whole given array

we loop from 0 to max digits as i variable
    and classify the elems in num array into 10 buckets based on their number on i th digit
    and concat every array from 10 buckets (preserve the order)
    then again do the same process until you reach max digits

*/

const GetDigit_at_i = (num, i) => {
  return Math.floor(Math.abs(num) / Math.pow(10, i)) % 10;
};

const GetDigits_of_Num = (num) => {
  if (num === 0) return 0;
  return Math.floor(Math.log10(num)) + 1;
};

const GetMax_Digits = (arr) => {
  let maxDigits = 0;
  for (let i = 0; i < arr.length; i++) {
    maxDigits = Math.max(GetDigits_of_Num(arr[i]), maxDigits);
  }
  return maxDigits;
};

const RadixSort = (arr) => {
  const maxDigits = GetMax_Digits(arr);
  for (let i = 0; i < maxDigits; i++) {
    //Create 10 buckets i.e arrays
    let digitBuckets = Array.from({ length: 10 }, () => []);
    for (let j = 0; j < arr.length; j++) {
      digitBuckets[GetDigit_at_i(arr[j], i)].push(arr[j]);
    }
    arr = [].concat(...digitBuckets);
  }
  return arr;
};

console.log(RadixSort([3445, 45, 3456, 45687, 35, 3, 2, 2, 78]));

File: Python/4_SortingAlgorithms/6_RadixSort.py
import math
def getDigitAtIndex(num, digit):
	return math.floor(abs(num) / math.pow(10, digit)) % 10;

def getNumDigits(num):
	if num == 0: return 0
	return math.floor(math.log10(num)) + 1

def getMaxNumDigits(lst):
	max_digits = -1
	for i in lst:
		max_digits = max(max_digits, getNumDigits(i))
	return max_digits

def RadixSort(lst):
	max_digits = getMaxNumDigits(lst)

	for i in range(max_digits):
		BUCKETS = [[], [], [], [], [], [], [], [], [], []]
		for value in lst:
			BUCKETS[getDigitAtIndex(value, i)].append(value)

		lst = []
		for bucket in BUCKETS:
			lst += bucket

	return lst

		


